_remake/lib/init-service-routes.js

```diff
@@ -0,0 +1,198 @@
+  else next();
+}
+
+export function initServiceRoutes({app}) {
+  // signup endpoint
+  // validation callbacks: validEmail, validPass
+  app.post('/service/signup', validEmail, validPass, (req, res) => {
+    const { email, password } = req.body;
+    const hashedPassword = bcrypt.hashSync(password, 8); // hash password
+    connection.query('INSERT INTO users (email, pwd_hash) VALUES ( ?, ?)',
+      [email, hashedPassword],
+      (err, results, fields) => {
+        if (err) {
+          if (err.code === "ER_DUP_ENTRY")
+            delete err.sql;
+          return res.status(500).json(err).end();
+        }
+        // generate JWT token based on user id and JWT_SECRET
+        const token = jwt.sign({ id: results.insertId }, config.jwt.secret, {
+          expiresIn: config.jwt.duration
+        });
+        return res.status(200).json({ token }).end();
+      });
+  })
+
+  // login endpoint
+  // validation callbacks: validEmail, validPass
+  app.post('/service/login', validEmail, validPass, (req, res) => {
+    const { email, password } = req.body;
+    connection.query('SELECT * FROM users WHERE email = ?',
+      [email],
+      (err, result, _) => {
+        if (err) return res.status(500).json(err).end();
+        const user = result[0];
+        // compare password with password hash
+        bcrypt.compare(password, user.pwd_hash, (err, passwordIsCorrect) => {
+          if (err) return res.status(500).json(err).end();
+          if (!passwordIsCorrect)
+            return res.status(403).json({ message: 'Wrong email or password' }).end();
+          // generate JWT token based on user id and JWT_SECRET
+          const token = jwt.sign({ id: user.id }, config.jwt.secret, {
+            expiresIn: config.jwt.duration
+          });
+          return res.status(200).json({ token }).end();
+        });
+      });
+  })
+
+  // endpoint for checking if subdomain is available (not already in the DB)
+  // validation callbacks: checkJWT, validSubdomain
+  // user must be authenticated to access it
+  app.get('/service/subdomain/check', checkJWT, validSubdomain, (req, res) => {
+    const { subdomain } = req.query;
+    connection.query('SELECT * FROM apps WHERE name = ?',
+      [subdomain],
+      (err, result, _) => {
+        if (err) return res.status(500).json(err).end();
+        if (result.length !== 0) return res.status(403).end();
+        return res.status(200).end();
+      });
+  });
+
+  // endpoint for registering subdomain (save in DB)
+  // validation callbacks: checkJWT, validSubdomain
+  // user must be authenticated to access it
+  app.post('/service/subdomain/register', checkJWT, validSubdomain, (req, res) => {
+    const { subdomain } = req.body;
+
+    connection.query('SELECT * FROM apps WHERE user_id = ?',
+      [req.user_id],
+      (err, results, fields) => {
+        if (err)
```

Comment:
In general, I'd prefer to wrap all statements in braces, even if there's just one line of code. This makes it much harder to mess up the code later when we might add another line to this statement.

Reply:
👍